### ISSUE TITLE
external "require('url')":1 Uncaught ReferenceError: require is not defined

### ISSUE THREAD

**Issue by okonon** [maximegris/angular-electron#60]
@maximegris 
When i run `npm run start:web` to serve the app from the browser i am getting following error.
It looks like chrome does not know how to interpret `require('url')` which is node module. the only place i can find this import is in webpack externals. Can you please point me into right direction?
```
external "require('url')":1 Uncaught ReferenceError: require is not defined
    at Object.<anonymous> (external "require('url')":1)
    at __webpack_require__ (bootstrap 151bbb78dee4aab80101:54)
    at Object.<anonymous> (client:2)
    at Object.<anonymous> (client?ffdb:204)
    at __webpack_require__ (bootstrap 151bbb78dee4aab80101:54)
    at Object.<anonymous> (polyfills.bundle.js:5881)
    at __webpack_require__ (bootstrap 151bbb78dee4aab80101:54)
    at webpackJsonpCallback (bootstrap 151bbb78dee4aab80101:25)
    at polyfills.bundle.js:1
14:57:33.753 core.es5.js:350 Uncaught reflect-metadata shim is required when using class decorators
```

**Comment by maximegris** [comment]
Hi @okonon
You may have import "URL" node modules even if you are in browser mode.
Look at `electron.service.ts` how node_module dependcies are loaded. (Conditional import with Typescript)

**Comment by kapit4n** [comment]
Sorry but where should we import the url and how sorry I dont know if this issue is because of some angular stuff or any library on electron.

**Comment by simboyy** [comment]
facing the same issue here

**Comment by okonon** [comment]
@maximegris i am getting this on newly cloned project. 
Steps to reproduce:
1. clone new project
2. npm install
3. npm run start:web

**Comment by dreamdevil00** [comment]
I found it lives in pollyfills.bundle.js  and the code 
```
/***/ "fs":
/***/ (function(module, exports) {

if(typeof require('fs') === 'undefined') {var e = new Error("Cannot find module \"require('fs')\""); e.code = 'MODULE_NOT_FOUND'; throw e;}
module.exports = require('fs');

/***/ }),

/***/ "url":
/***/ (function(module, exports) {

module.exports = require('url');

/***/ })
```

As you can see,  when importing the url module, it doesn't check if it exists before. So why?

**Comment by nikulinsanya** [comment]
The same issue I can not understand how to solve. Is it something in webpack configuration?

**Comment by maximegris** [comment]
What version of this repo do you use ? (I suggest to migrate to v3.x. There were a lot of refactoring)
Do you import node/electron modules in your code ?

**Comment by AngryPowman** [comment]
I have the same issue, it seems like I had ejected the project

**Comment by maximegris** [comment]
All node modules import are done in `electron.service.ts` ? And all use of node modules are wrapped by an if(this.electronService.isElectron()) to be sure you are using it in an electron app and not a webapp?

**Comment by mkotsollaris** [comment]
Facing same issue here..

**Comment by mkotsollaris** [comment]
I managed to solve this by adding `target: 'web'` in my `webpack.config.js` and restarting my application with `npm start`!

**Comment by maximegris** [comment]
Please have a look at the last version (4.1.0) [especially commits  165c50a1e964718cbde9881db230a134194438eb 1c403a80fad3be355ed99d924f4d8601a34aa117 50657d01265ec5c9dddc957e3d3e5288b1220753]
It should makes node modules (and other native libs) working and remove require errors.

**Comment by simonmicheal** [comment]
this is still happening

**Comment by whyboris** [comment]
I had this problem happen recently when upgrading from Electron 4 to Electron 5.

*Solution*: https://stackoverflow.com/questions/55093700/electron-5-0-0-uncaught-referenceerror-require-is-not-defined

> It turns out, nodeIntegration was true by default in previous electron versions, but false by default in 5.0.0.

Setting `nodeIntegration` to `true` will may fix your upgrade from Electron 4 to 5:
```
new BrowserWindow({
    webPreferences: {
      nodeIntegration: true,
```

**Comment by zombodotcom** [comment]
doesn't work for me =(

**Comment by thomascayne** [comment]
@maximegris Great project. Thank you. This is my 3rd project with it.

I just cloned `angular-electron` to start a new project (3rd). The same error with **require** came up with `npm start` for _only_ `http://localhost:4200/` (as in my 2nd project which I never spent time to figure out. Now I did). `npm start` invokes both web & electron app. Which is the ideal scenario.

Also, running `ng:serve` produces the **require** error for _only_ `http://localhost:4200/`.

*The solution*: 

- Do **not** use `ng:serve`
- Do **use**: `ng:serve:web`

**or** change `npm start` to:

`"start": "npm-run-all -p ng:serve:web electron:serve"`

I suggest removing reference to `ng:serve`.

**Comment by charlesr1971** [comment]
@thomascayne 

Hi Thomas

I am getting:

`Uncaught ReferenceError: require is not defined`

I want to try and use your solution by adding the following to _package.json -> scripts_:

`"start": "npm-run-all -p ng:serve:web electron:serve"`

But I cannot see what **ng:serve:web** is an alias for?

When I try:

`npm run start`

I get:

```
ERROR: Task not found: "ng:serve:web"
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! twitter-circle-app-angular-v4@3.0.0 start: `npm-run-all -p ng:serve:web electron:serve`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the twitter-circle-app-angular-v4@3.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

When I use:

`npm run electron:local`

Everything works perfectly, but I cannot hot reload, so it is quite tiresome, working within Chromium

**Comment by thomascayne** [comment]
@charlesr1971 It's being a while since I worked with this repo. I don't have it on any of my systems. I would recommended following @maximegris solution at [Changing nodeIntegration to false breaks entire app](https://github.com/maximegris/angular-electron/issues/508#issuecomment-673058049).

**Comment by charlesr1971** [comment]
@thomascayne Thanks Thomas, but when I swap to:

`"start": "npm-run-all -p ng:serve:web electron:serve"`

And do:

`npm run start`

I get this error:

```
ERROR: Task not found: "ng:serve:web"
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! twitter-circle-app-angular-v4@3.0.0 start: `npm-run-all -p ng:serve:web electron:serve`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the twitter-circle-app-angular-v4@3.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

**Comment by thomascayne** [comment]
The script you quoted above is no longer referenced in version 10.4.1. Notice is it now separated into it's own script:

`"ng:serve": "ng serve -c web -o"`

Try removing `ng:serve:web`  from `"start": "npm-run-all -p ng:serve:web electron:serve"`

**Comment by charlesr1971** [comment]
@thomascayne So you mean:

`"start": "npm-run-all -p electron:serve",`

**Comment by charlesr1971** [comment]
Here is my _package.json_

```
{
  "name": "twitter-circle-app-angular-v4",
  "version": "3.0.0",
  "description": "Angular 12 with Electron (Typescript + SASS + Hot Reload)",
  "homepage": "https://github.com/charlesr1971/twitter-circle-app-angular-v4",
  "author": {
    "name": "Establish Mindfulness Ltd",
    "email": "admin@establishmindfulness.com"
  },
  "keywords": [
    "angular",
    "angular 12",
    "electron",
    "nodejs",
    "typescript",
    "spectron",
    "eslint",
    "sass",
    "windows",
    "mac",
    "linux"
  ],
  "main": "app/main.js",
  "private": true,
  "scripts": {
    "postinstall": "electron-builder install-app-deps",
    "ng": "ng",
    "start": "npm-run-all -p electron:serve",
    "build": "npm run electron:serve-tsc && ng build --base-href ./",
    "build:dev": "npm run build -- -c dev",
    "build:prod": "npm run build -- -c production",
    "ng:serve": "ng serve -c web -o",
    "electron:serve-tsc": "tsc -p tsconfig.serve.json",
    "electron:serve": "wait-on tcp:4200 && npm run electron:serve-tsc && npx electron . --serve",
    "electron:local": "npm run build:prod && npx electron .",
    "electron:build": "npm run build:prod && electron-builder build --publish=never",
    "test": "ng test --watch=false",
    "test:watch": "ng test",
    "e2e": "npm run build:prod && cross-env TS_NODE_PROJECT='e2e/tsconfig.e2e.json' mocha --timeout 300000 --require ts-node/register e2e/**/*.e2e.ts",
    "version": "conventional-changelog -i CHANGELOG.md -s -r 0 && git add CHANGELOG.md",
    "lint": "ng lint"
  },
  "dependencies": {
    "@angular/animations": "12.1.2",
    "@angular/cdk": "12.2.13",
    "@angular/common": "12.1.2",
    "@angular/compiler": "12.1.2",
    "@angular/core": "12.1.2",
    "@angular/forms": "12.1.2",
    "@angular/language-service": "12.1.2",
    "@angular/material": "12.2.13",
    "@angular/platform-browser": "12.1.2",
    "@angular/platform-browser-dynamic": "12.1.2",
    "@angular/router": "12.1.2",
    "@electron/remote": "1.2.0",
    "bitmap-ts": "1.0.3",
    "findit": "2.0.0",
    "https": "1.0.0",
    "image-size": "1.0.0",
    "ngx-colors": "2.10.1",
    "pureimage": "0.3.5",
    "rxjs": "~6.6.0",
    "tslib": "^2.1.0",
    "twitter-lite": "1.1.0",
    "twitter-lite-custom": "file:app/twitter-lite-custom",
    "zone.js": "~0.11.4"
  },
  "devDependencies": {
    "@angular-builders/custom-webpack": "12.1.0",
    "@angular-devkit/build-angular": "12.1.2",
    "@angular-eslint/builder": "12.3.1",
    "@angular-eslint/eslint-plugin": "12.3.1",
    "@angular-eslint/eslint-plugin-template": "12.3.1",
    "@angular-eslint/schematics": "12.3.1",
    "@angular-eslint/template-parser": "12.3.1",
    "@angular/cli": "12.1.2",
    "@angular/compiler-cli": "12.1.2",
    "@ngx-translate/core": "13.0.0",
    "@ngx-translate/http-loader": "6.0.0",
    "@types/jasmine": "3.8.1",
    "@types/jasminewd2": "2.0.10",
    "@types/mocha": "8.2.3",
    "@types/node": "16.3.3",
    "@typescript-eslint/eslint-plugin": "4.28.3",
    "@typescript-eslint/parser": "4.28.3",
    "chai": "4.3.4",
    "conventional-changelog-cli": "2.1.1",
    "cross-env": "7.0.3",
    "electron": "13.1.7",
    "electron-builder": "22.10.5",
    "electron-reload": "1.5.0",
    "eslint": "7.30.0",
    "eslint-plugin-import": "2.23.4",
    "eslint-plugin-jsdoc": "35.4.3",
    "eslint-plugin-prefer-arrow": "1.2.3",
    "jasmine-core": "3.8.0",
    "jasmine-spec-reporter": "7.0.0",
    "karma": "6.3.4",
    "karma-coverage-istanbul-reporter": "3.0.3",
    "karma-electron": "7.0.0",
    "karma-jasmine": "4.0.1",
    "karma-jasmine-html-reporter": "1.7.0",
    "mocha": "9.0.2",
    "nan": "2.14.2",
    "npm-run-all": "4.1.5",
    "spectron": "15.0.0",
    "ts-node": "10.1.0",
    "typescript": "~4.2.0",
    "wait-on": "5.0.1",
    "webdriver-manager": "12.1.8"
  },
  "engines": {
    "node": ">=14.0.0"
  },
  "browserslist": [
    "chrome 91"
  ]
}

```

**Comment by charlesr1971** [comment]
@thomascayne Hi Thomas
In the end, I decided to download the latest version of _Angular Electron 10.4.1_
The problem is, I was missing the NodeJs plugins for _WebKit 5+_, and the latest version has fixed this issue. 
It took me all day to set the project up again. It’s a shame there isn’t a way to update the **Angular Electron** core? I decided to clone the maximegris repo, install my node modules on top and then add my code back in, gradually.
But, amazingly, I managed to get everything working again, but this time with Angular 13 & Electron 16. 
I have noticed, it loads about twice as quickly. 
So, there have been some major improvements made, under the hood! 🙂
The main thing is that I can now access the application via the browser using **localhost:4200** 🙏
